### PR TITLE
Finish the initial Configuration functional API and added unit test

### PR DESCRIPTION
**Pull request by apsoto**
- Added Configuration method to support setting options
- Added Configuration Unit Test
- Renamed Config to Option since it was confusing to have Config and Configuration

### CODE REVIEW

**Comment by apsoto** [comment]
I thought the intent was to do something like:

    config.WithDebug(true).WithApplicationId("foo")

I couldn't find anyway to make the `WithXX()` api to work without adding the missing method and switching it to

    config.Option(WithDebug(true), WithApplicationID(ConfigTestID))

@asciifaceman LMK if I butchered your design ;)

**Comment by asciifaceman** [comment]
You are chaining wrong (I've done research to determine if what  you're example above works and it doesn't really).

ex.
```
striketracker.NewClientWithOptions(
    striketracker.WithParameter(),
    striketracker.WithOtherParameter(),
)
```

They are functional overloads essentially

This could be imported as a first class citizen of course and become

```
NewClientWithOptions(
    WithParameter(),
    WithOtherParameter(),
)
```

But that is an end-user implementation decision (and open to conflicts)

File: striketracker/config.go [hunk 1]
```diff
@@ -21,25 +21,33 @@ func (c *Configuration) Validate() error {
 	return nil
 }
 
-// Config is a functional API for configuring the client
-type Config func(*Configuration)
+// Option is a functional API for configuring the client.
+// https://commandcenter.blogspot.com/2014/01/self-referential-functions-and-design.html
+type Option func(*Configuration)
```
Comment (asciifaceman) [inline striketracker/config.go:38]: This is fine, I'm still in that phase of development where things are getting renamed a lot and this is definitely  more appropriate

**Comment by asciifaceman** [comment]
Here is a Dave Cheney example of the functional option pattern (Dave is a golang god and  core  language contributor - he's made decisions on how go works)

https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

This may be a better comment reference than the link you have

File: striketracker/config_test.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+)
+
+const (
+	ConfigTestToken = "98fg798hu7sd9hu32kj23kj23"
+	ConfigTestID    = "TestApplication"
+)
+
+func TestFunctionalApi(t *testing.T) {
+	config := Configuration{}
+	config.Option(WithDebug(true), WithApplicationID(ConfigTestID), WithAuthorizationHeaderToken(ConfigTestToken))
```
Comment (asciifaceman) [inline striketracker/config_test.go]: This is actually not a bad pass I think without your implementing client.

I'll have to verify later and play around a bit. The client implementation looks like this (as I'm sure you've seen)

```
func NewClientWithOptions(opts ...Config) (*Client, error) {
	options := &Configuration{}
	for _, opt := range opts {
		opt(options)
	}
```

Which if I'm not mistaken, translates roughly to what you are doing

Comment (apsoto) [reply striketracker/config_test.go]: I added `NewConfiguration` and updated the unit tests

File: striketracker/config_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		t.Errorf("Expected App Id to be empty")
+	}
+
+	if config.AuthorizationHeaderToken != "" {
+		t.Errorf("Expected token to be empty")
+	}
+}
+
+func TestSettingOptions(t *testing.T) {
+	config, err := NewConfiguration(WithDebug(true), WithApplicationID(ConfigTestID), WithAuthorizationHeaderToken(ConfigTestToken))
```
Comment (asciifaceman) [inline striketracker/config_test.go]: Little more readable this way but not required as I'll just fix it later lol

```
config, err := NewConfiguration(
    WithDebug(true),
    WithApplicationID(ConfigTestID),
    WithAuthorizationHeaderToken(ConfigTestToken),
)
```

File: striketracker/config_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+	if err != nil {
+		t.Errorf("Did not expect error")
```
Comment (asciifaceman) [inline striketracker/config_test.go]: "Did not expect error but received %v", err

also for the others you should show what you expected and what you got (even if its obvious).

```
	if config.ApplicationID != ConfigTestID {
		t.Errorf("Expected ApplicationId to be %s but got %s", ConfigTestID, config.ApplicationID)
	}
```

Comment (apsoto) [reply striketracker/config_test.go]: yea, I was being lazy, also I'm used test assert functions that do that for me ;)

**Comment by asciifaceman** [comment]
@apsoto sorry to throw more on your plate, feel free to toss this back at me but:

Could you add a `NewClientFromConfiguration` method in client.go, which takes in the Config struck and creates the client similar to `NewClientWithOptions`?

I can provide any snippets you would need to do this, and I think with that this PR would be done (just need something to use the NewConfiguration with, as another instantiation option - there may be cases where an end user is, say, JSON-managing configuration direct into the struct and want to just pass it right in)

**Pull request merged**